### PR TITLE
[Merged by Bors] - feat(Topology/Algebra/InfiniteSum/Real): Add partition lemma

### PR DESCRIPTION
**Pull request by CBirkbeck**
We also  `summable_partition` (and the required  `Set.sigmaEquiv`), needed for #10377.  

----------

Ignore the typo in the PR name! 

- [x] depends on: #12503 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Algebra/InfiniteSum/Real.lean [hunk 1]
```diff
@@ -94,3 +61,61 @@ theorem dist_le_tsum_dist_of_tendsto₀ (h : Summable fun n ↦ dist (f n) (f n.
...
+  rw [← not_iff_not, Classical.not_not, not_summable_iff_tendsto_nat_atTop_of_nonneg hf]
+#align summable_iff_not_tendsto_nat_at_top_of_nonneg summable_iff_not_tendsto_nat_atTop_of_nonneg
+
+theorem summable_sigma_of_nonneg {β : α → Type*} {f : (Σ x, β x) → ℝ} (hf : ∀ x, 0 ≤ f x) :
+    Summable f ↔ (∀ x, Summable fun y => f ⟨x, y⟩) ∧ Summable fun x => ∑' y, f ⟨x, y⟩ := by
+  lift f to (Σx, β x) → ℝ≥0 using hf
+  exact mod_cast NNReal.summable_sigma
+#align summable_sigma_of_nonneg summable_sigma_of_nonneg
+
+lemma summable_partition {κ ι: Type*} {f : ι → ℝ} (hf : 0 ≤ f) {s : κ → Set ι}
```
Comment (grunweg) [inline Mathlib/Topology/Algebra/InfiniteSum/Real.lean:84]: For future reviewers: this is the only new lemma; all others are just moved.

File: Mathlib/Topology/Instances/ENNReal.lean [hunk 2]
```diff
@@ -1497,6 +1449,27 @@ theorem Filter.Tendsto.edist {f g : β → α} {x : Filter β} {a b : α} (hf :
   (continuous_edist.tendsto (a, b)).comp (hf.prod_mk_nhds hg)
 #align filter.tendsto.edist Filter.Tendsto.edist
 
+/-- If the extended distance between consecutive points of a sequence is estimated
+by a summable series of `NNReal`s, then the original sequence is a Cauchy sequence. -/
+theorem cauchySeq_of_edist_le_of_summable [PseudoEMetricSpace α] {f : ℕ → α} (d : ℕ → ℝ≥0)
+    (hf : ∀ n, edist (f n) (f n.succ) ≤ d n) (hd : Summable d) : CauchySeq f := by
+  refine EMetric.cauchySeq_iff_NNReal.2 fun ε εpos ↦ ?_
+  -- Actually we need partial sums of `d` to be a Cauchy sequence
```
Comment (grunweg) [inline Mathlib/Topology/Instances/ENNReal.lean:1457]: Pre-existing: I would put a full stop at the end of each comment, as these are complete sentences.

**Review by grunweg — APPROVED** [review]
I can confirm that this PR is almost only code motion, replacing a proof by a commented more elegant version and two new declarations.

This PR would be easier to review if you split the move from the new code: I approve the code motion/porting TODO parts (but didn't scrutinize the others carefully).

**Comment by CBirkbeck** [comment]
OK I split this into two PRs. The code moving is now #12503. Once that goes in this will only have the two new results.

**Comment by adomani** [comment]
Again, the `no_lost_declarations` script reports

* `+` `lemma summable_partition {κ ι: Type*} {f : ι → ℝ} (hf : 0 ≤ f) {s : κ → Set ι}`

---

* 1  added declarations
* 0  removed declarations
* 9  paired declarations

---

**Comment by loefflerd** [comment]
This needs a rebase now #12503 is in, please feel free to ping me for a review once you've merged master

File: Mathlib/Topology/Algebra/InfiniteSum/Real.lean [hunk 3]
```diff
@@ -81,6 +81,13 @@ theorem summable_sigma_of_nonneg {β : α → Type*} {f : (Σ x, β x) → ℝ}
   exact mod_cast NNReal.summable_sigma
 #align summable_sigma_of_nonneg summable_sigma_of_nonneg
 
+lemma summable_partition {κ ι: Type*} {f : ι → ℝ} (hf : 0 ≤ f) {s : κ → Set ι}
+    (hs : ∀ i, ∃! j, i ∈ s j) : Summable f ↔
+      (∀ j, Summable fun i : s j ↦ f i) ∧ Summable fun j ↦ ∑' i : s j, f i := by
+  rw [← (Set.sigmaEquiv s hs).summable_iff, summable_sigma_of_nonneg]
+  simp only [Set.sigmaEquiv, Equiv.coe_fn_mk, Function.comp_apply]
+  exact fun _ ↦ hf _
```
Comment (loefflerd) [inline Mathlib/Topology/Algebra/InfiniteSum/Real.lean:89]:
```suggestion
  simpa only [← (Set.sigmaEquiv s hs).summable_iff] using summable_sigma_of_nonneg (fun _ ↦ hf _)
```

File: Mathlib/Topology/Algebra/InfiniteSum/Real.lean [hunk 4]
```diff
@@ -81,6 +81,13 @@ theorem summable_sigma_of_nonneg {β : α → Type*} {f : (Σ x, β x) → ℝ}
   exact mod_cast NNReal.summable_sigma
 #align summable_sigma_of_nonneg summable_sigma_of_nonneg
 
+lemma summable_partition {κ ι: Type*} {f : ι → ℝ} (hf : 0 ≤ f) {s : κ → Set ι}
```
Comment (loefflerd) [inline Mathlib/Topology/Algebra/InfiniteSum/Real.lean:84]:
```suggestion
lemma summable_partition {κ ι : Type*} {f : ι → ℝ} (hf : 0 ≤ f) {s : κ → Set ι}
```
(Maybe also rename the types consistently with the above lemma)

File: Mathlib/Data/Set/Lattice.lean [hunk 5]
```diff
@@ -2197,6 +2197,14 @@ theorem sigmaToiUnion_bijective (h : Pairwise fun i j => Disjoint (t i) (t j)) :
   ⟨sigmaToiUnion_injective t h, sigmaToiUnion_surjective t⟩
 #align set.sigma_to_Union_bijective Set.sigmaToiUnion_bijective
 
+/-- Equivalence between the sigma of a family of sets of `β` and `β`. -/
+noncomputable def sigmaEquiv {ι κ : Type*} (s : κ → Set ι) (hs : ∀ i, ∃! j, i ∈ s j) :
+    (Σ j, s j) ≃ ι where
+  toFun x := x.2
+  invFun x := ⟨(hs x).choose, x, (hs x).choose_spec.1⟩
+  left_inv x := by ext; exacts [((hs x.2).choose_spec.2 x.1 x.2.2).symm, rfl]
+  right_inv x := by rfl
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply Mathlib/Data/Set/Lattice.lean]: Yes thank you, this def/lemma went through lots of versions and I forgot to make the description and the notation better!

**Comment by riccardobrasca** [comment]
This is very small and I think we can just merge it, but if you start using partition a lot (meaning `(s : κ → Set ι) (hs : ∀ i, ∃! j, i ∈ s j`) we probably want a small API, but we will see. Thanks!

bors merge

**Comment by CBirkbeck** [comment]
> This is very small and I think we can just merge it, but if you start using partition a lot (meaning `(s : κ → Set ι) (hs : ∀ i, ∃! j, i ∈ s j`) we probably want a small API, but we will see. Thanks!
> 

Yeah I don't know how much it'll be used for example for the stuff Yael wanted the Finest.box for. But yes we can add more API if needed.

**Pull request closed without merging**